File: puppyscript/cpp-puppy/macro_rules.cpp
//
// Created by impodog on 7/30/2023.
//
#include "processor.h"

std::regex RE_MACRO_DEF(R"(\s*macro\s+(\w+(!?))\s*:\s*)"),
        RE_FORM_DEF(R"(\s*form(\s+(\w[\w\s.]*))?\s*:\s*)");

using FormArgs = std::unordered_map<std::string, std::string>;

class Form {
public:
    string_list args, body;

    Form(const std::string &args, const std::string &body) {
        this->args = split(args, ' ');
        auto base_indent = indent_of(body);
        auto lines = split(body, '\n');
        for (const auto &line: lines) {
            this->body.push_back(line.substr(base_indent.size()));
        }
    }

    string_list replace_with(const FormArgs &assign, const std::string &base_indent) const {
        string_list result;
        bool is_first = true;
        for (auto line: body) {
            for (const auto &repl: assign) {
                while (true) {
                    size_t pos = line.find(repl.first);
                    if (pos == std::string::npos) {
                        break;
                    }
                    size_t end = pos + repl.first.size();
                    if (pos > 0 && line[pos - 1] == '#') {
                        pos--;
                    }
                    if (end < line.size() && line[end] == '#') {
                        end++;
                    }
                    line = line.substr(0, pos) + repl.second + line.substr(end);
                }
            }
            if (is_first) {
                result.push_back(line);
                is_first = false;
            } else {
                result.push_back(base_indent + line);
            }
        }
        return result;
    }
};

class Macro {
public:
    std::string name;
    std::vector<Form> forms;

    Macro(std::string name, std::vector<Form> forms) :
            name(std::move(name)), forms(std::move(forms)) {

    }

    std::string try_replace_with(const std::string &args_str, const std::string &base_indent) const {
        auto args = split(args_str, ' ');
        size_t arg_len = args.size();
        for (const auto &form: forms) {
            FormArgs assign;
            size_t form_len = form.args.size();
            if (form_len > arg_len) {
                continue;
            }
            if (form_len == 0) {
                if (arg_len == 0) {
                    // intended to do nothing
                } else {
                    continue;
                }
            } else if (ends_with(form.args.back(), ELLIPSIS)) {
                for (size_t i = 0; i < form_len - 1; i++) {
                    assign.insert({form.args[i], args[i]});
                }
                const auto &e_name = form.args.back();
                assign[e_name.substr(0, e_name.size() - ELLIPSIS_SIZE)] =
                        join(args.begin() + form_len - 1, args.end(), ' ');
            } else if (ends_with(form.args.front(), ELLIPSIS)) {
                for (size_t i = 1; i < form_len; i++) {
                    assign.insert({form.args[i], args[arg_len - form_len + i]});
                }
                const auto &e_name = form.args.front();
                assign[e_name.substr(0, e_name.size() - ELLIPSIS_SIZE)] =
                        join(args.begin(), args.end() - form_len + 1, ' ');
            } else if (arg_len == form_len) {
                for (size_t i = 0; i < form_len; i++) {
                    assign.insert({form.args[i], args[i]});
                }
            }

            if (assign.size() == form_len) {
                return join(form.replace_with(assign, base_indent), '\n');
            }
        }
        throw std::runtime_error("");
    }
};

class MacroRules {
    std::string content;
    std::vector<Macro> macros;
    bool changed;

    std::string replace(std::string part, std::string base_indent) {
        if (part.empty() || is_spaces(part)) {
            return "";
        }
        for (const auto &macro: macros) {
            std::smatch results;
            if (std::regex_search(part, results, std::regex(macro.name + R"(\<)"))) {
                size_t index = results.length();
                size_t depth = 1;
                while (depth) {
                    if (index == part.size()) {
                        throw std::runtime_error("No matching '>' found in line: " + part);
                    }
                    switch (part[index]) {
                        case '<':
                            depth++;
                            break;
                        case '>':
                            depth--;
                            break;
                        default:
                            break;
                    }
                    index++;
                }
                auto args = replace(part.substr(results.length(), index - results.length() - 1), "");
                try {
                    part = results.prefix().str() +
                           macro.try_replace_with(args, base_indent) +
                           part.substr(index);
                    changed = true;
                } catch (std::runtime_error &e) {}
            }
        }
        return part;
    }

    void scan_macro_defs() {
        string_list result;
        std::string name, macro_outer, form_args, form_outer, form_body;
        std::string cur_indent, line;
        std::vector<Form> forms;
        bool macro_started = false, form_started = false;

        const auto test_next_macro = [&result, &name, &macro_outer, &cur_indent, &line, &macro_started]() {
            std::smatch results;
            if (std::regex_search(line, results, RE_MACRO_DEF)) {
                name = results[1].str();
                macro_started = true;
                if (name.back() != '!') {
                    std::cerr << "Macro names should end with '!'. "
                              << "Consider changing \"" << name << "\" to \""
                              << name << "\"!" << std::endl;
                }
                macro_outer = cur_indent;
            } else {
                result.push_back(line);
                macro_started = false;
            }
        };

        const auto add_form = [&form_args, &form_body, &forms, &form_started]() {
            forms.emplace_back(form_args, form_body);
            form_started = false;
        };

        const auto test_next_form = [&form_outer, &form_args, &form_body, &form_started, &cur_indent, &line]() {
            std::smatch results;
            if (std::regex_search(line, results, RE_FORM_DEF)) {
                form_outer = cur_indent;
                if (results[1].str().empty()) {
                    form_args = "";
                } else {
                    form_args = results[2].str();
                }
                form_body.clear();
                form_started = true;
            } else {
                if (!is_debug_info(line)) {
                    std::cerr << "Stray line in macro definition: \""
                              << line << "\"" << std::endl;
                }
            }
        };

        auto lines = split(content, '\n');

        for (const auto &line_: lines) {
            line = line_;
            cur_indent = indent_of(line);
            if (!macro_started) {
                test_next_macro();
            } else {
                if (form_started) {
                    if (cur_indent.size() == form_outer.size()) {
                        add_form();
                        test_next_form();
                    } else if (cur_indent.size() <= form_outer.size()) {
                        add_form();
                        macros.emplace_back(name, forms);
                        macro_started = form_started = false;
                        forms.clear();
                        test_next_macro();
                    } else if (!is_debug_info(line)) {
                        form_body.append(line).push_back('\n');
                    }
                } else {
                    test_next_form();
                }
            }
        }
        content = join(result, '\n');
    }

    void replace_all() {
        changed = true;
        while (changed) {
            string_list result;
            changed = false;
            auto lines = split(content, '\n');
            for (auto line: lines) {
                line = replace(line, indent_of(line));
                result.push_back(line);
            }
            content = join(result, '\n');
        }
    }

public:
    MacroRules(const path &file) {
        content = read(file);
    }

    path work(const std::string &output_name) {
        scan_macro_defs();
        replace_all();

        path output = output_name + ".macro_rules.puppy";
        write(output, content);
        return output;
    }
};

PROCESSOR(macro_rules) {
    MacroRules macro_rules(file);
    return macro_rules.work(output_name);
}

File: puppyscript/puppy/macro_rules.py
import re
import warnings

from .ids import firsts, is_debug_info


def ellipsis_name_re(name: str):
    return name.strip(".").replace(".", r"\.")


class Form:
    args: list[str]
    body: list[str]

    def __init__(self, args: str, body: str):
        self.args = list()
        for arg in args.split(" "):
            if len(arg) > 0 and not arg.isspace():
                self.args.append(arg)
        base_indent  = len(firsts(body))
        self.body = list()
        for line in body.split("\n"):
            self.body.append(line[base_indent:])

    def repl_with(self, assign: dict[str, str], base_indent: str) -> list[str]:
        result = list()
        is_first = True
        for line in self.body:
            for name, value in assign.items():
                tmp = re.search(r"(\b|#)%s(\b|#)" % name, line)
                if tmp is not None:
                    start = tmp.end(1)
                    end = tmp.start(2)
                    if tmp.group(1) == "#":
                        start -= 1
                    if tmp.group(2) == "#":
                        end += 1
                    line = line[:start] + value + line[end:]

            if is_first:
                result.append(line)
            else:
                result.append(base_indent + line)
            is_first = False
        return result


class Macro:
    name: str
    forms: list[Form]

    def __init__(self, name: str, forms: list[Form]):
        self.name = name
        self.forms = forms.copy()

    def try_repl(self, args_str: str, base_indent: str) -> str:
        args = list()
        for arg in args_str.split(" "):
            if len(arg) > 0 and not arg.isspace():
                args.append(arg.strip())
        arg_len = len(args)
        for form in self.forms:
            assign: dict[str, str] = dict()
            form_len = len(form.args)
            if form_len > arg_len:
                continue
            if form_len == 0:
                if arg_len == 0:
                    ...
                else:
                    continue
            elif form.args[-1].endswith("..."):
                for i in range(form_len - 1):
                    assign[form.args[i]] = args[i]
                assign[ellipsis_name_re(form.args[-1])] = " ".join(args[form_len - 1:])
            elif form.args[0].endswith("..."):
                for i in range(1, form_len):
                    assign[form.args[i]] = args[arg_len - form_len + i]
                assign[ellipsis_name_re(form.args[0])] = " ".join(args[:arg_len - form_len + 1])
            elif arg_len == form_len:
                for i in range(arg_len):
                    assign[form.args[i]] = args[i]

            if len(assign) == form_len:
                return "\n".join(form.repl_with(assign, base_indent))
        raise ValueError


class MacroRules:
    content: str
    macros: list[Macro]
    success: bool

    def __init__(self, file: str):
        with open(file, "r", encoding="utf-8") as f:
            self.content = f.read()
        self.macros = list()

    def repl(self, part: str, base_indent: str) -> str:
        if len(part) == 0 or part.isspace():
            return ""
        for macro in self.macros:
            tmp = re.search("%s<" % macro.name, part)
            if tmp is None:
                continue
            index = tmp.end()
            depth = 1
            while depth:
                match part[index]:
                    case "<":
                        depth += 1
                    case ">":
                        depth -= 1
                index += 1
            args = self.repl(part[tmp.end():index - 1], "")
            try:
                part = part[:tmp.start()] + macro.try_repl(args, base_indent) + part[index:]
            except:
                continue
            self.success = True
        return part

    def scan_macro_defs(self):
        result = list()
        name: str | None = None
        macro_outer: str | None = None
        form_args: str | None = None
        form_outer: str | None = None
        form_body: list[str] | None = None
        forms: list[Form] = list()

        def test_next_macro():
            nonlocal result, name, macro_outer
            tmp = re.fullmatch(r"\s*macro\s+(\w+(!?))\s*:\s*", line)
            if tmp is None:
                result.append(line)
            else:
                name = tmp.group(1)
                if not name.endswith("!"):
                    warnings.warn(
                        UserWarning("Macro names should end with \"!\". Consider changing \"%s\" to \"%s\"." % (
                            name, name + "!"))
                    )
                macro_outer = cur_indent

        def add_form():
            nonlocal form_args, form_body
            forms.append(Form(form_args, "\n".join(form_body)))
            form_args = None

        def start_form():
            nonlocal form_outer, form_args, form_body
            tmp = re.fullmatch(r"\s*form(\s+(\w[\w\s.]*))?\s*:\s*", line)
            if tmp is None:
                if not is_debug_info(line):
                    warnings.warn(
                        UserWarning("Stray line in macro forms: \"%s\"." % line)
                    )
            else:
                form_outer = cur_indent
                if tmp.group(1) is None:
                    form_args = ""
                else:
                    form_args = tmp.group(1)
                form_body = list()

        for line in self.content.split("\n"):
            cur_indent = firsts(line)
            if name is None:
                test_next_macro()
            else:
                if form_args is None:
                    start_form()
                else:
                    if len(cur_indent) == len(form_outer):
                        add_form()
                        start_form()
                    elif len(cur_indent) <= len(macro_outer):
                        add_form()
                        self.macros.append(Macro(name, forms))
                        name = form_args = None
                        forms.clear()
                        test_next_macro()
                    elif not is_debug_info(line):
                        form_body.append(line)
        self.content = "\n".join(result)

    def replace_all(self):
        self.success = True
        while self.success:
            result = list()
            self.success = False
            for line in self.content.split("\n"):
                line = self.repl(line, firsts(line))
                result.append(line)
            self.content = "\n".join(result)

    def work(self, output_name: str) -> str:
        self.scan_macro_defs()
        self.replace_all()
        output = output_name + ".macros"
        with open(output, "w", encoding="utf-8") as f:
            f.write(self.content)
        return output
